### DISCUSSION TITLE
"Literal[1]" is incompatible with "SupportsIndex" -- bug?

### DISCUSSION THREAD

**Discussion by mjperrone** [microsoft/pyright#6023] [General]
Here is the code:

```python
endpoint = "api-http"
endpoint.replace("api-", "", 1)
```

Here is the error:
```
file.py:2:30 - error: Argument of type "Literal[1]" cannot be assigned to parameter "__count" of type "SupportsIndex | _ProtocolMeta" in function "replace"
    Type "Literal[1]" cannot be assigned to type "SupportsIndex | _ProtocolMeta"
      "Literal[1]" is incompatible with "SupportsIndex"
      "Literal[1]" is incompatible with "_ProtocolMeta" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

referring to the last argument `1`.

Here is the definition of replace I could find:
https://github.com/python/typeshed/blob/main/stdlib/builtins.pyi#L512-L515

```python
    @overload
    def replace(
        self: LiteralString, __old: LiteralString, __new: LiteralString, __count: SupportsIndex = -1
    ) -> LiteralString: ...
```

It is unclear to me if this is an issue with with typeshed definitions or pyright, but it seems like _something_ is wrong for `1` not to be acceptable to give to `replace`.

**Comment by erictraut** [comment]
Thanks for reporting. I noticed a similar issue a few days ago and have already introduced a fix. It's related to overload matching, and it's in the gray area between a bug and "intended behavior". Since overload matching behavior is not fully spec'ed and already uses some heuristics, I introduce another heuristic into the mix. Your code sample now type checks without error. This will be included in tomorrow's release of pyright (1.1.329).

**Reply by mjperrone** [reply]
```bash
$ pyright test.py
/Users/mperrone/code/gabledata/product/test.py
  /Users/mperrone/code/gabledata/product/test.py:2:30 - error: Argument of type "Literal[1]" cannot be assigned to parameter "__count" of type "SupportsIndex | _ProtocolMeta" in function "replace"
    Type "Literal[1]" cannot be assigned to type "SupportsIndex | _ProtocolMeta"
      "Literal[1]" is incompatible with "SupportsIndex"
      "Literal[1]" is incompatible with "_ProtocolMeta" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
$ pyright --version
pyright 1.1.329
```

@erictraut, did it actually get released in 1.1.329? I am still seeing this behavior.

**Reply by erictraut** [reply]
I'm not able to repro this error. Let's see if we can figure out what's different between your configuration and mine. Which version of Python are you using? Which platform? Are you able to repro the problem with a fresh venv? Did you install pyright via the community-maintained Python wrapper from pypi, or did you use npm to install it?

**Reply by mjperrone** [reply]
I am using python `3.10.11` on mac os. I can repro this with a fresh venv. I am using poetry to install the community maintained python wrapper from pipy. I updated my `pyproject.toml` to include `pyright = "1.1.329"` and then ran `poetry install`. Then:

```bash
$ poetry run pyright --version
pyright 1.1.329
$ poetry run pyright test.py
```
and I get the error I showed above. Some more context for you:

```bash
poetry show pyright       
 name         : pyright                          
 version      : 1.1.329                          
 description  : Command line wrapper for pyright 

dependencies
 - nodeenv >=1.6.0
 
$ poetry run which pyright  
/Users/mperrone/code/my/repo/.venv/bin/pyright

$ poetry run python --version
Python 3.10.11

$ echo $OSTYPE
darwin22.0
```

**Reply by mjperrone** [reply]
Can you point to the pull request or the commit that you think addresses this problem?

**Reply by erictraut** [reply]
I've heard of poetry but have never used it. Perhaps that's what's interfering with pyright. Could you try to repro without poetry?

**Reply by erictraut** [reply]
If you would like me to try to repro with poetry, please provide additional instructions, including any configuration files you're using.

**Reply by mjperrone** [reply]
Well here is something you should be able to copy and paste to repro the poetry setup assuming you have `pip` and `python` already:

```bash
# official poetry install method: https://python-poetry.org/docs/
curl -sSL https://install.python-poetry.org/ | python3 -
cat > pyproject.toml <<EOF
[tool.poetry]
name = "showpyright"
version = "0.1.0"
description = ""
authors = ["Your Name <[you@example.com](mailto:you@example.com)>"]
readme = "README.md"

[tool.poetry.dependencies]
python = "^3.11"

[tool.poetry.group.dev.dependencies]
pyright = "1.1.329"

[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"
EOF
poetry install
cat > test.py <<EOF
endpoint = "api-http"
endpoint.replace("api-", "", 1)
EOF
poetry run pyright test.py
```

However, when I do this in a new project, I don't get any errors. So there's something up with this particular project that's causing the difference. I will dig into that a little more

**Reply by mjperrone** [reply]
Hmm, it is something to do with the virtualenvironment that poetry is installing into. If I don't activate that project environment and use my global environment to install that version of pyright and use that to test the file in my special project, no errors.

But the `--version` is the same for both :|

**Reply by mjperrone** [reply]
I wanted to eliminate the possibility that this was an issue with the python package, so I installed `pyright` directly with `yarn`.

It is very strange to me that this python development tool is not officially published in the python ecosystem, and I don't find it acceptable to require devs to install node to develop(using pyright) in a python project. However, since this problem is happening to me in a polyglot (python+typescript) monorepo, I already have a node environment set up, so it's not an _extra_ burden in this case.

Unfortunately, using the node version does not yield different results :(

```bash
$ yarn run which pyright
yarn run v1.22.19
/Users/mperrone/code/my/repo/node_modules/.bin/pyright
$ yarn run pyright --version
yarn run v1.22.19
pyright 1.1.329
yarn run pyright test.py
  /Users/mperrone/code/gabledata/product/test.py:2:30 - error: Argument of type "Literal[1]" cannot be assigned to parameter "__count" of type "SupportsIndex | _ProtocolMeta" in function "replace"
    Type "Literal[1]" cannot be assigned to type "SupportsIndex | _ProtocolMeta"
      "Literal[1]" is incompatible with "SupportsIndex"
      "Literal[1]" is incompatible with "_ProtocolMeta" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

My `pyrightconfig.json` in this project:

```json
{
  "exclude": [
    "**/node_modules",
    "**/__pycache__",
    "**/.venv",
    "**/dist"
  ]
}
```

pyproject.toml pyright section:

```toml
[tool.pyright]
include = ["src"]
exclude = ["**/node_modules",
    "**/__pycache__",
    "**/.venv"
]
```

**Reply by mjperrone** [reply]
To summarize:

* Using `poetry` to install the python wrapper of `pyright` in this project yields this `Literal[1]` error
* Using `yarn` to install `pyright` in this project yields the `Literal[1]` error
* Using `pip` in a global virtual env to install the python wrapper of `pyright` in this project yields no errors

I see that `--verbose` shows 

```
Loading configuration file at /Users/mperrone/code/gabledata/product/pyrightconfig.json
Assuming Python version 3.10
```

But that's just _where_ it got the config, and that's the same with all three executables. Is there a way to get `pyright` to report exactly what configuration options it is using?

**Reply by erictraut** [reply]
I don't think it has anything to do with configuration options in pyright. I can't think of any configuration options that would cause this error to appear.

Based on your analysis, it appears the issue is linked to your virtual environment and its contents, so let's explore that further. How was your venv created? Is it built from a distinct list of installed dependencies (e.g. a `requirements.txt` file), or is it (like many of my own venvs) the result of ad hoc package installations over time? If it's built from a list of installed dependencies, are you able to create a new venv from scratch that results in the same behavior? If so, could you send me that list of dependencies so I could try to repro on my system?

**Reply by mjperrone** [reply]
> Based on your analysis, it appears the issue is linked to your virtual environment and its contents

I think this is correct. I have another variable that I haven't reported that I think may help understand the difference: osx terminal vs VSCode integrated terminal. The integrated terminal automatically `source`s the venv and yields the error, but osx terminal does not.

When I `source .venv/bin/activate` in the osx terminal, then the failure happens. I think this confirms your interpretation.

> How was your venv created?
`poetry` handles the `.venv` creation on `poetry install`.

> Is it built from a distinct list of installed dependencies?

Yes, much like `yarn` in `node`, package version constraints are are specified in `pyproject.toml`, then `poetry` resolves those constraints to a specific set of packages, which is recorded in `poetry.lock` (like `yarn.lock`). Then poetry internally uses `pip` to  install into the virtual environment.

> If it's built from a list of installed dependencies, are you able to create a new venv from scratch that results in the same behavior? If so, could you send me that list of dependencies so I could try to repro on my system?


I tried to do this, but when I do, the error doesn't happen 🤦 . I'm going to try another way to work around this.
